numpy/core/tests/test_multiarray.py

```diff
@@ -9191,3 +9191,54 @@ def test_getfield():
     pytest.raises(ValueError, a.getfield, 'uint8', -1)
     pytest.raises(ValueError, a.getfield, 'uint8', 16)
     pytest.raises(ValueError, a.getfield, 'uint64', 0)
+
+def TestViewDtype():
+    """
+    Verify that making a view of a non-contiguous array works as expected.
+    """
+    def test_smaller_dtype_multiple(self):
+        x = np.arange(10, dtype='<i4')[::2]  # Non-contiguous
+        pytest.raises(ValueError, x.view, '<i2')
+        expected = np.array([[0, 0], [0, 2], [0, 4], [0, 6], [0, 8]], '<i2')
+        assert_array_equal(x[:, None].view('<i2'), expected)
+
+    def test_smaller_dtype_not_multiple(self):
+        x = np.arange(5, dtype='<i4')[::2]  # Non-contiguous
+        pytest.raises(ValueError, x.view, 'S3')
+        pytest.raises(ValueError, x[:, None].view, 'S3')
+        # Make sure the problem is because of the dtype size
+        expected = [[b''], [b'\x00\x00\x02'], [b'\x00\x00\x04'],
+                    [b'\x00\x00\x06'], [b'\x00\x00\x08']]
+        assert_array_equal(x[:, None].view('S4'), expected)
+
+    def test_larger_dtype_multiple(self):
+        x = np.arange(20, dtype='<i2').reshape(10, 2)[::2]  # Non-contiguous
+        expected = np.array([[65536], [327684], [589832],
+                             [851980], [1114128]], dtype='<i4')
+        assert_array_equal(x.view('<i4'), expected)
+
+    def test_larger_dtype_not_multiple(self):
+        x = np.arange(20, dtype='<i2').reshape(10, 2)[::2, :]
+        pytest.raises(ValueError, x.view, 'S3')
+        # Make sure the problem is because of the dtype size
+        expected = [[b'\x00\x00\x01'], [b'\x04\x00\x05'], [b'\x08\x00\t'],
+                    [b'\x0c\x00\r'], [b'\x10\x00\x11']]
+        assert_array_equal(x.view('S4'), expected)
+
+    def test_unit_axis(self):
+        """
+        Test that last unit axis is selected
+        """
+        x = np.arange(6 * 8 * 10, dtype='<i4').reshape(6, 8, 10)
+        x = x[::2, None, ::2, None, ::2]
+        assert x.view('<i2').shape == (3, 1, 4, 2, 5)
+
+    def test_contract_axis(self):
+        """
+        Test that last contiguous axis is selected.
+        """
+        x = np.arange(3 * 4 * 5, dtype='<i2').reshape(3, 4, 5)
+        x = np.lib.stride_stricks.as_strided(x,
+            strides=(x.strides[0], x.strides[2], x.strides[2], x.strides[1]),
+            shape=(x.shape[0], 2, x.shape[2] - 1, x.shape[1]))
+        assert x.view('<i4').shape == (3, 2, 2, 4)
```

Comment:
Could you add some tests that check the failure modes?

Reply:
Do you have specific ones in mind? I'm checking the ones I introduced already, and the existing ones are presumably exercised elsewhere.

Reply:
looking at the "Files" tab on the PR, I see that Coverage is complaining about some of the error code paths.

Reply:
I see what the problem was now. I checked the new conditions, but never verified that the original was still being raised. All done now.